opennmt/utils/decoding.py

```diff
@@ -172,18 +172,24 @@ def initialize(self, batch_size, start_ids):
     initial_log_probs = tf.tile([0.] + [-float("inf")] * (self.beam_size - 1), [batch_size])
     parent_ids = tf.TensorArray(tf.int32, size=0, dynamic_size=True)
     maximum_lengths = tf.zeros([batch_size], dtype=tf.int32)
-    return start_ids, finished, initial_log_probs, (parent_ids, maximum_lengths)
+    sequence_lengths = tf.zeros([batch_size * self.beam_size], dtype=tf.int32)
+    return start_ids, finished, initial_log_probs, (parent_ids, maximum_lengths, sequence_lengths)
 
   def step(self, step, sampler, log_probs, cum_log_probs, finished, state, extra_vars):
-    parent_ids, maximum_lengths = extra_vars
+    parent_ids, maximum_lengths, sequence_lengths = extra_vars
 
     # Compute scores from log probabilities.
     vocab_size = log_probs.shape.as_list()[-1]
     total_probs = log_probs + tf.expand_dims(cum_log_probs, 1)  # Add current beam probability.
     total_probs = tf.reshape(total_probs, [-1, self.beam_size * vocab_size])
     scores = total_probs
     if self.length_penalty != 0:
-      scores /= tf.pow(((5. + tf.cast(step + 1, tf.float32)) / 6.), self.length_penalty)
+      expand_sequence_lengths = tf.tile(tf.expand_dims(sequence_lengths, 1),
```

Comment:
To avoid the tiling, the `scores` could be reshaped after applying the length penalty. `total_probs` does not need to be reshaped:

```python
total_probs = log_probs + tf.expand_dims(cum_log_probs, 1)  # Add current beam probability.
scores = total_probs
if self.length_penalty != 0:
  ...
scores = tf.reshape(scores, [-1, self.beam_size * vocab_size])
```

Reply:
My bad, `total_probs` should also be reshaped. So both reshape can occur after the length penalty.